packages/core/src/lib/markdown-it/index.ts

```diff
@@ -144,14 +146,33 @@ markdownIt.renderer.rules.fence = (tokens: Token[],
   // wrap all lines with <span> so we can number them
   str = lines.map((line, index) => {
     const currentLineNumber = index + 1;
-    const rule = highlightRules.find(highlightRule =>
-      highlightRule.shouldApplyHighlight(currentLineNumber));
-    if (rule) {
-      return rule.applyHighlight(line, currentLineNumber);
+    // Rules that affects this line
+    const rules = highlightRules.filter(
+      highlightRule => highlightRule.shouldApplyHighlight(currentLineNumber));
+    if (rules.length === 0) {
+      // not highlighted
+      return `<span>${line}\n</span>`;
+    }
+    // FIXME: Reason why two rules cannot be applied on the same line
+    const rawBoundaries: Boundary[] = [];
+    let lineHighlightType = HIGHLIGHT_TYPES.PartialText;
+    // Per line WholeLine override WholeText overrides PartialText
+    for (let i = 0; i < rules.length; i += 1) {
+      const { highlightType, boundaries } = rules[i].getHighlightType(currentLineNumber);
+      if (highlightType === HIGHLIGHT_TYPES.WholeLine) {
+        lineHighlightType = HIGHLIGHT_TYPES.WholeLine;
+        return Highlighter.highlightWholeLine(line);
+      } else if (highlightType === HIGHLIGHT_TYPES.WholeText) {
+        lineHighlightType = HIGHLIGHT_TYPES.WholeText;
+      } else if (boundaries !== null) {
+        boundaries.forEach(boundary => rawBoundaries.push(boundary));
+      }
     }
```

Comment:
Can we simplify this loop slightly to make it more directly represent that it's looking for overrides?
e.g. if any of the rules is WholeLine, return immediately without entering the loop
```suggestion
    // WholeLine overrides all
    if (rules.some(rule => rule.getHighlightType(currentLineNumber).highlightType === HIGHLIGHT_TYPES.WholeLine)) {
        return Highlighter.highlightWholeLine(line);
    } 

    // WholeText overrides PartialText
    if (rules.some(rule => rule.getHighlightType(currentLineNumber).highlightType === HIGHLIGHT_TYPES.WholeText)) {
      return Highlighter.highlightWholeText(line); 
    }

    rules.forEach(rule => {
      const { highlightType, boundaries } = rule.getHighlightType(currentLineNumber);
      if (highlightType === HIGHLIGHT_TYPES.PartialText && boundaries !== null) {
        boundaries.forEach(boundary => rawBoundaries.push(boundary));
      }
    })
    return Highlighter.highlightPartOfText(line, rawBoundaries);
```

Reply:
Don't really have an issue with this but will require three passes with this right? Certainly better readability

Reply:
Noted on the performance concern with 3 passes - I think performance should be fine with three passes since the size of the highlight rules array is seldom very large

Reply:
> Noted on the performance concern with 3 passes - I think performance should be fine with three passes since the size of the highlight rules array is seldom very large

Ok I will make the change accordingly then

Reply:
> Ok I will make the change accordingly then

Just checking: has it been made?

Reply:
@zlimez please make the change accordingly, almost had a stroke reading this code 🫠

Reply:
upon review, we can keep the code change, but some helpful inline comments would be necessary to understand why the code is written this way.